lib/timeline_records.rb

```diff
@@ -0,0 +1,298 @@
+      class_imports.where.not(id: patient_imports) if patient_imports.present?
+    {
+      class_imports:
+        class_imports
+          .group_by(&:session_id)
+          .transform_values { |imports| imports.map(&:id) },
+      cohort_imports:
+        patient
+          .organisation
+          .cohort_imports
+          .map(&:id)
+          .reject do |ci|
+            patient_events(patient)[:cohort_imports].include?(ci.id)
+          end
+    }
+  end
+
+  def patient_events(patient)
+    {
+      class_imports: patient.class_imports.map(&:id),
+      cohort_imports: patient.cohort_imports.map(&:id),
+      sessions: patient.sessions.map(&:id)
+    }
+  end
+
+  def extract_detail_config(detail_config)
+    detail_config.deep_symbolize_keys
+  end
+
+  def details
+    @details ||= DEFAULT_DETAILS_CONFIG.merge(@detail_config)
+  end
+
+  def audits
+    @audits ||= DEFAULT_AUDITS_CONFIG.merge(@audit_config)
+  end
+
+  def load_events(event_names)
+    event_names.each do |event_name|
+      event_type = event_name.to_sym
+
+      if details.key?(event_type)
+        fields = details[event_type]
+        records = @patient.send(event_type)
+        records = Array(records)
+
+        records.each do |record|
+          event_details =
+            fields.each_with_object({}) do |field, hash|
+              field_value = record.send(field)
+              hash[field.to_s] = field_value.nil? ? "nil" : field_value
+            end
+          @events << {
+            event_type: record.class.name,
+            id: record.id,
+            details: event_details,
+            created_at: record.created_at
+          }
+        end
+      else
+        custom_event_handler(event_type)
+      end
+    end
+    @events.sort_by! { |event| event[:created_at] }.reverse!
+  end
+
+  def load_grouped_events(event_names)
+    load_events(event_names)
+    @events.each { |event| event[:details] = format_details(event) }
+    @grouped_events_by_date =
+      @events.group_by do |event|
+        event[:created_at].strftime(Date::DATE_FORMATS[:long])
+      end
+    @grouped_events =
+      @grouped_events_by_date.sort_by { |date, _events| date }.reverse!.to_h
+  end
+
+  private
+
+  def format_details(event)
+    if event[:details].is_a?(Hash)
+      event[:details].map { |k, v| "#{k}: #{v}" }.join(", ")
+    else
+      event[:details].to_s
+    end
+  end
+
+  def custom_event_handler(event_type)
+    case event_type
+    when :org_cohort_imports
+      @events += org_cohort_imports_events
+    when /^add_class_imports_\d+$/ # e.g. add_class_imports_123
+      session_id = event_type.to_s.split("_").last
+      @events += add_class_imports_events(session_id)
+    when :audits
+      @events += audits_events
+    else
+      puts "No handler for event type: #{event_type}"
+    end
+  end
+
+  def org_cohort_imports_events
+    @additional_events[:cohort_imports].map do |cohort_import_id|
+      cohort_import = CohortImport.find(cohort_import_id)
```

Comment:
```suggestion
    CohortImport.where(id: @additional_events[:cohort_imports]).map do |cohort_import|
```

I would suggest doing something like this (and in other similar places) to make only 1 query to the database and fetch all the imports, rather than needing to make a query to the database for each one.